README.md
# KastleX - Kafka REST Proxy in Elixir
Kastle is a REST interface to Kafka cluster, powered by [Brod](https://github.com/klarna/brod) and [Phoenix framework](http://www.phoenixframework.org/).

See also [Kastle](https://github.com/klarna/kastle).

# Get started

    mix deps.get
    mix phx.server

To start with an interactive shell:

    iex --sname kastlex -S mix phx.server

To start an interactive shell with all applications paths loaded but not starting any application

    iex -S mix shell

By default KastleX will try to connect to kafka at localhost:9092 and to zookeeper on localhost:2181.

Default app port is 8092.

## Tests

    docker-compose -f docker-compose-test.yml up -d
    # give it ~10 seconds
    # force create __consumer_offsets
    docker exec -it kafka /opt/kafka/bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic kastlex
    # Ctrl+C to terminate consumer process
    mix test

KastleX expects to find kafka:9092 and zookeeper:2181 on localhost.
It also needs `kastlex` topic with a single partition, and
`auto.create.topics.enable` parameter in Kafka's server.properties set
to `false`.

## Manual testing

1. Bring environment up
    ```
        docker-compose up
    ```
2. Start consumer group
    ```
        kafkacat -b localhost:9092 -G test-consumer-group test-topic
    ```
3. Start producing to that consumer group
    ```
        LC_CTYPE=C watch -n 1 "cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1 | kafkacat -P -b localhost:9092 -t test-topic -p 0"
    ```
4. Access Prometheus [http://localhost:9090](http://localhost:9090) and verify that metrics were consumed.

# API

## Produce messages

    POST /api/v1/messages/:topic
    POST /api/v1/messages/:topic/:partition

With the first version KastleX will pick a random partition to produce message to.  
Use `Content-type: application/binary`.  
Key is supplied as query parameter `key`.  
Value is request body.  
Successful response should have:
- HTTP Status 204
- HTTP header "x-kafka-partition" set for which partition the message was produced to
- HTTP header "x-message-offset" set for which offset the message was persisted at
- Empty HTTP body

cURL example (-d implies POST):

    curl -s -i -H "Content-Type: application/binary" localhost:8092/api/v1/messages/foo -d bar

## Fetch messages

### v1

GET /api/v1/messages/:topic/:partition
GET /api/v1/messages/:topic/:partition/:offset

Body

```
{
  "size": null,
  "messages": [
    {
      "value": "test-data",
      "ts_type": "create",
      "ts": 1536157879011,
      "offset": 45,
      "key": null,
      "headers": {
        "foo": "1",
        "baz": "null",
        "bar": "s"
      }
    }
  ],
  "highWmOffset": 46,
  "error_code": "no_error"
}
```

### v2

GET /api/v2/messages/:topic/:partition
GET /api/v2/messages/:topic/:partition/:offset

Header
```
x-high-wm-offset: 46
```

Body

```
[
  {
    "value": "test-data",
    "ts_type": "create",
    "ts": 1536157879011,
    "offset": 45,
    "key": null,
    "headers": {
      "foo": "1",
      "baz": "null",
      "bar": "s"
    }
  }
]
```

If `offset` is not given the `last` message is fetched by default.

Optional parameters:
  * `max_wait_time`: maximum time in ms to wait for the response, default 1000
  * `min_bytes`: minimum bytes to accumulate in the response, default 1
  * `max_bytes`: maximum bytes to fetch, default 100 kB

`offset` can be an exact offset, `last`, `latest`, `earliest` or negative.
Default `last`.
When negative, KastleX will assume it's a relative offset to the `latest`.
i.e. `-1` is the same as `last`

With `Accept: application/json` header the response will include all
of the messages returned from kafka with their metadata

With `Accept: application/binary` header KastleX will return the value
in http response body. Message offset, headers, ts_type and ts are put to http header.
An example from cURL output:

```
$ curl -v -H "Accept: application/binary" localhost:8092/api/v2/messages/kastlex/0

< x-request-id: 2l8sgeskov545ica5c000005
< content-type: application/binary; charset=utf-8
< x-high-wm-offset: 654
< x-message-headers: {"foo":"1","baz":"null","bar":"s"}
< x-message-offset: 653
< x-message-ts: 1536228953642
< x-message-ts-type: :create
```

## Query available offsets for partition.

    GET /api/v1/offsets/:topic/:partition
    {"offset": 20}

Optional parameters:
  * `at`: point of interest, `latest`, `earliest`, or a number, default `latest`

## Consumer groups

    GET /api/v1/consumers
    ["console-consumer-25992"]

    GET /api/v1/consumers/:group_id
    {
        "protocol": "range",
        "partitions": [
            {
                "topic": "kastlex",
                "partition": 0,
                "offset": 20,
                "metadata_encoding": "text",
                "metadata": "",
                "high_wm_offset": 20,
                "expire_time": 1473714215481,
                "commit_time": 1473627815481
            }
        ],
        "members": [
            {
                "subscription": {
                    "version": 0,
                    "user_data_encoding": "text",
                    "user_data": "",
                    "topics": [
                        "kastlex"
                    ]
                },
                "session_timeout": 30000,
                "member_id": "consumer-1-ea5aa1bc-6b14-488f-88f1-26edb2261786",
                "client_id": "consumer-1",
                "client_host": "/127.0.0.1",
                "assignment": {
                    "version": 0,
                    "user_data_encoding": "text",
                    "user_data": "",
                    "topic_partitions": {
                        "kastlex": [
                            0
                        ]
                    }
                }
            }
        ],
        "leader": "consumer-1-ea5aa1bc-6b14-488f-88f1-26edb2261786",
        "group_id": "console-consumer-66960",
        "generation_id": 1
    }


## Topics metadata

    GET /api/v1/topics
    ["kastlex"]

    GET /api/v1/topics/:topic
    {"topic":"kastlex","partitions":[{"replicas":[0],"partition":0,"leader":0,"isr":[0]}],"config":{}}

## Brokers metadata

    GET /api/v1/brokers
    [{"port":9092,"id":0,"host":"localhost","endpoints":["PLAINTEXT://127.0.0.1:9092"]}]

    GET /api/v1/brokers/:broker_id
    {"port":9092,"id":0,"host":"localhost","endpoints":["PLAINTEXT://127.0.0.1:9092"]}

(Yes, this one looks a bit silly)

## List under-replicated partitions

    GET /api/v1/urp
    GET /api/v1/urp/:topic

# Authentication
Authentication is a courtesy of [Guardian](https://github.com/ueberauth/guardian).

There are 2 files, permissions.yml and passwd.yml to configure permissions for different actions.

Example permissions.yml:

    anonymous:
      list_topics: true
      show_topic: all
      list_brokers: true
      show_broker: all
      show_offsets: all
      fetch: all
      list_urps: true
      show_urps: all
      list_groups: true
      show_group: all
    user1:
      produce:
        - kastlex
    admin:
      reload: true
      revoke: true

Anonymous user can do pretty much everything except writing data to kafka.

`user` can write to topic `kastlex`.

`admin` can reload permissions.

`all` means access to all topics, replace it with a list of specific topics when applicable (see for example user.produce).

Example passwd.yml:

    user1:
      password_hash: "$2b$12$3iR64t7Sm.cAHtZs5jkxZehdWQ7knmN/NxmK.X7NBUHfiIAxT4T9y"
    admin:
      password_hash: "$2b$12$gp5pJc/AGclJradJC9DuHe6xJoIe5HOwtAUGe2z7QFeAjvw1eZUKW"

Here we specify password hashes for each user.

`user` has password `user`, `admin` has password `admin`. Simple.

Generate password:

    mix hashpw difficult2guess

To obtain a token users need to login (submit a form with 2 fields, username and password):

    curl localhost:8092/login --data "username=user1&password=user1"
    {"token":"bcrypt hash"}

Get token directly into a shell variable (requires `jq`):

    JWT=$(curl -s localhost:8092/login --data "username=user1&password=user1" | jq -r .token)

Then you can submit authenticated requests via curl as:

    curl -H "Authorization: Bearer $JWT" localhost:8092/admin/reload
    curl -H "Authorization: Bearer $JWT" localhost:8092/api/v1/messages/kastlex/0 -H "Content-type: application/binary" -d 1

# Token storage
It is possible to run KastleX in 2 modes: when tokens are issued with relatively short timespan and administrator does not have any control over them, and when tokens are long-lived, but persisted and can be revoked on demand.

KastleX is using a compacted topic (cleanup_policy=compact) with a single partition in Kafka as a token storage. In order to enable token storage, add the following hook config for Guardian application:

    hooks: Kastlex.TokenStorage

And configuration for TokenStorage:

    config :kastlex, Kastlex.TokenStorage,
      topic: "_kastlex_tokens"

Alternatively set the following environment variable:

    KASTLEX_ENABLE_TOKEN_STORAGE=1

These 2 can be used to alter default storage topic name and default ttl:

    KASTLEX_TOKEN_STORAGE_TOPIC=_kastlex_tokens
    KASTLEX_TOKEN_TTL_SECONDS=315360000

Administrator can use the following API endpoint to revoke a token:

    DELETE  /admin/tokens/:username

Correspinding permission item in permissions.yml file is 'revoke'.

# Deployment to production

## Generate a JWK

    openssl ecparam -name secp521r1 -genkey -noout -out jwk.pem

## Generate a secret key base

    printf "%s" $(openssl rand -base64 32 | tr -d =) > secret_base.key

## Set the following varibles for Kastlex environment

    KASTLEX_SECRET_KEY_BASE=/path/to/secret_base.key
    KASTLEX_JWK_FILE=/path/to/jwk.pem
    KASTLEX_PERMISSIONS_FILE_PATH=/path/to/permissions.yml
    KASTLEX_PASSWD_FILE_PATH=/path/to/passwd.yml
    KASTLEX_KAFKA_CLUSTER=kafka-host1:9092,kafka-host2:9092

## (Optional) Set custom HTTP port

    KASTLEX_HTTP_PORT=8092

## (Optional) Enable HTTPS
Variables are given with their default values except for KASTLEX_USE_HTTPS which is disabled by default.

So if you just set `KASTLEX_USE_HTTPS=true`, Kastlex will be accepting TLS connection on 8093 and use certificates in /etc/kastlex/ssl.

    KASTLEX_USE_HTTPS=true
    KASTLEX_HTTPS_PORT=8093
    KASTLEX_CERTFILE=/etc/kastlex/ssl/server.crt
    KASTLEX_KEYFILE=/etc/kastlex/ssl/server.key
    KASTLEX_CACERTFILE=/etc/kastlex/ssl/ca-cert.crt

## More optional variables

    KASTLEX_KAFKA_USE_SSL=false
    KASTLEX_KAFKA_CACERTFILE=/path/to/cacertfile
    KASTLEX_KAFKA_CERTFILE=/path/to/certfile
    KASTLEX_KAFKA_KEYFILE=/path/to/keyfile
    KASTLEX_KAFKA_SASL_FILE=/path/to/file/with/sasl/credentials
    KASTLEX_PRODUCER_REQUIRED_ACKS=
    KASTLEX_PRODUCER_ACK_TIMEOUT=
    KASTLEX_PRODUCER_PARTITION_BUFFER_LIMIT=
    KASTLEX_PRODUCER_PARTITION_ONWIRE_LIMIT=
    KASTLEX_PRODUCER_MAX_BATCH_SIZE=
    KASTLEX_PRODUCER_MAX_RETRIES=
    KASTLEX_PRODUCER_RETRY_BACKOFF_MS=
    KASTLEX_PRODUCER_MAX_LINGER_MS=
    KASTLEX_PRODUCER_MAX_LINGER_COUNT=

File with sasl credentials is a plain text yml like file having `username`, `password` and `mechanism`,
where `mechanism` is optional which supports `plain` (default), `scram_sha_256` or `scram_sha_512`.
For example:

```
username: user
password: s3cr3t
mechanism: scram_sha_512
```

If the variable is set, and file exists, KastleX will use SASL authentication when connecting to Kafka.

## Building release for production

    MIX_ENV=prod mix compile
    MIX_ENV=prod mix release

## Running release

    rel/kastlex/bin/kastlex console


changelog.md
  - Delete unused config `cg_cache_dir` since cg state is no longer cached in dets
  - Reverted changelog.md from newest-last to newest-first

* 1.7.4
  - Remove expire_time from GUI since kafka 2.0 does not support it any more

* 1.7.3
  - Upgrade to brod 3.7.11 to fix a bug when receiving empty batch from compacted transactions

* 1.7.2
  - Fix prod config, change allowed algos for guardian app from HS512 to ES512
    also upgrade to brod 3.7.10 to fix a bug which may skip through unstable messages
    in case messages are published in a transaction
* 1.7.1
  - Change dets to ets for consumer state cache (port from 1.5.6).

* 1.7.0
  - Deleted zookeeper client, fetch topic metadata from KAFKA APIs
  - Added version 2 'consumers' API, fetch consumer states from KAFKA APIs

* 1.6.0
  - Upgrade to elixir 1.7
  - Upgrade to brod 3.7
  - Add scram-sasl auth towards kafka
  - Error codes changed (per change in kafka spec and convention) from CamelCase to snake_case, e.g:
      * UnknownTopicOrPartition -> unknown_topic_or_partition
      * LeaderNotAvailable -> leader_not_available
  - Message fileds `crc`, `attributes` and `magic_byte` are removed from JSON fetch response
  - Per kafka message `headers` field is added to JSON fetch response.
  - New http headers `x-message-headers`, `x-message-offset`, `x-message-ts` and `x-message-ts-type` are
    added to both v1 and v2 binary fetch response.
    The header value is the kafka message header formated in JSON.
  - Add `x-kafka-partition` and `x-message-offset` http headers to produce response
  - Stopped using 'latest as last' in fetch APIs, instead, introduced new logical offset 'last'.
    NOTE: in case the partition is empty, fetching 'last' will result in error.

* 1.5.6 Change consumer cg cache from dets to ets (detached from master, proted to 1.7.1)

* 1.5.5 Fix return result of Kastlex.Users.get_user

* 1.5.4 Upgrade to brod 3.3.4.
  Prior to 3.3.4, brod consumer group coordinator commits consumed offsets to kafka, instead of next-to-fetch offset.
  Which is different from the convention in kafka spec and other library implementations.
  For the same reason, KastleX calculated lagging as `high-watermark-offset - committed-offset - 1`,
  1.5.4 changed it to `high-watermark-offset - committed-offset`.

* 1.5.3 Improve Prometheus metrics report performance


lib/kastlex.ex
defmodule Kastlex do
  use Application

  require Logger

  def start(_type, _args) do
    import Supervisor.Spec, warn: false

    # http endpoint
    endpoint = Application.fetch_env!(:kastlex, Kastlex.Endpoint)
    http = endpoint[:http]
    port = system_env("KASTLEX_HTTP_PORT", http[:port])
    Logger.info "HTTP port: #{port}"
    http = Keyword.put(http, :port, port)
    Application.put_env(:kastlex, Kastlex.Endpoint, Keyword.put(endpoint, :http, http))
    maybe_init_https(system_env("KASTLEX_USE_HTTPS", false, &s2b/1))
    maybe_set_secret_key_base(system_env("KASTLEX_SECRET_KEY_BASE"))
    maybe_set_guardian_secret_key(system_env("KASTLEX_JWK_FILE"))

    # authentication/authorization
    permission_file_default_path = Path.join(File.cwd!(), "permissions.yml")
    permissions_file_path = system_env("KASTLEX_PERMISSIONS_FILE_PATH", permission_file_default_path)
    Logger.info "Permissions file path: #{permissions_file_path}"
    Application.put_env(:kastlex, :permissions_file_path, permissions_file_path)
    passwd_file_default_path = Path.join(File.cwd!(), "passwd.yml")
    passwd_file_path = system_env("KASTLEX_PASSWD_FILE_PATH", passwd_file_default_path)
    Logger.info "Passwd file path: #{passwd_file_path}"
    Application.put_env(:kastlex, :passwd_file_path, passwd_file_path)

    # consumer groups
    cg_exclude_regex = system_env("KASTLEX_CG_EXCLUDE_REGEX", nil)
    maybe_log_parameter("Consumer groups exclude regexp", cg_exclude_regex)
    Application.put_env(:kastlex, :cg_exclude_regex, cg_exclude_regex)

    # token storage
    maybe_configure_token_storage(system_env("KASTLEX_ENABLE_TOKEN_STORAGE", false, &s2b/1))
    maybe_set_token_ttl(system_env("KASTLEX_TOKEN_TTL_SECONDS"))

    # kafka connection
    kafka_cluster = system_env("KASTLEX_KAFKA_CLUSTER", [{'localhost', 9092}], &parse_endpoints/1)
    Logger.info "Kafka cluster: #{inspect kafka_cluster}"

    brod_client_config = get_brod_client_config(true)
    :ok = :brod.start_client(kafka_cluster, :kastlex, brod_client_config)

    children = [
      worker(Kastlex.Users, []),
      worker(Kastlex.TokenStorage, []),
      supervisor(Kastlex.Collectors, []),
      supervisor(Kastlex.Endpoint, []),
      supervisor(Phoenix.PubSub.PG2, [Kastlex.PubSub, []]),
    ]

    opts = [strategy: :one_for_one, name: Kastlex.Supervisor]
    Supervisor.start_link(children, opts)
  end

  def get_brod_client_id(), do: :kastlex

  # Tell Phoenix to update the endpoint configuration
  # whenever the application is updated.
  def config_change(changed, _new, removed) do
    Kastlex.Endpoint.config_change(changed, removed)
    :ok
  end

  def get_user(name), do: Kastlex.Users.get_user(name)

  def get_anonymous(), do: Kastlex.Users.get_anonymous()

  def reload() do
    Kastlex.Users.reload()
  end

  def token_storage_enabled?() do
    guardian = Application.fetch_env!(:guardian, Guardian)
    guardian[:hooks] == Kastlex.TokenStorage
  end

  defp maybe_init_https(true) do
    Logger.info "Using HTTPS"
    port = system_env("KASTLEX_HTTPS_PORT", 8093)
    Logger.info "HTTPS port: #{port}"
    keyfile = system_env("KASTLEX_KEYFILE", "/etc/kastlex/ssl/server.key")
    Logger.info "Keyfile: #{keyfile}"
    certfile = system_env("KASTLEX_CERTFILE", "/etc/kastlex/ssl/server.crt")
    Logger.info "certfile: #{certfile}"
    cacertfile = system_env("KASTLEX_CACERTFILE", "/etc/kastlex/ssl/ca-cert.crt")
    Logger.info "cacertfile: #{cacertfile}"
    config = [port: port, keyfile: keyfile, certfile: certfile, cacertfile: cacertfile]
    endpoint = Application.fetch_env!(:kastlex, Kastlex.Endpoint)
    Application.put_env(:kastlex, Kastlex.Endpoint, Keyword.put(endpoint, :https, config))
  end
  defp maybe_init_https(_), do: :ok

  defp maybe_set_secret_key_base(nil), do: :ok
  defp maybe_set_secret_key_base(secret_key_base) do
    Logger.info "Using custom secret key base from file: #{secret_key_base}"
    endpoint = Application.fetch_env!(:kastlex, Kastlex.Endpoint)
    Application.put_env(:kastlex, Kastlex.Endpoint,
                        Keyword.put(endpoint, :secret_key_base, secret_key_base))
  end

  defp maybe_set_guardian_secret_key(nil) do
    init_secret_key(Application.get_env(:guardian, Guardian)[:secret_key_file])
  end
  defp maybe_set_guardian_secret_key(file) do
    init_secret_key(file)
  end

  defp init_secret_key(file) do
    Logger.info "Reading secret key from #{file}"
    jwk = JOSE.JWK.from_pem_file(file)
    guardian = Application.fetch_env!(:guardian, Guardian)
    Application.put_env(:guardian, Guardian,
                        Keyword.put(guardian, :secret_key, jwk))
  end

  defp maybe_configure_token_storage(true) do
    Logger.info "OS env override: enabling token storage"
    guardian = Application.fetch_env!(:guardian, Guardian)
    Application.put_env(:guardian, Guardian,
                        Keyword.put(guardian, :hooks, Kastlex.TokenStorage))
    case System.get_env("KASTLEX_TOKEN_STORAGE_TOPIC") do
      nil -> :ok
      topic ->
        Application.put_env(:kastlex, Kastlex.TokenStorage, [topic: topic])
        Logger.info "Custom token storage topic: #{topic}"
    end
  end
  defp maybe_configure_token_storage(_), do: :ok

  defp maybe_set_token_ttl(nil), do: :ok
  defp maybe_set_token_ttl(ttl) do
    Logger.info "Using custom token ttl: #{ttl} seconds"
    {ttl, _} = Integer.parse(ttl)
    guardian = Application.fetch_env!(:guardian, Guardian)
    Application.put_env(:guardian, Guardian,
                        Keyword.put(guardian, :ttl, {ttl, :seconds}))
  end

  def get_brod_client_config(maybe_log \\ false) do
    log = case maybe_log do
      true -> fn(x) -> Logger.info(x) end
      false -> fn(_) -> :ok end
    end
    ssl_config =
    []
      |> maybe_put(:cacertfile, system_env("KASTLEX_KAFKA_CACERTFILE"))
      |> maybe_put(:certfile, system_env("KASTLEX_KAFKA_CERTFILE"))
      |> maybe_put(:keyfile, system_env("KASTLEX_KAFKA_KEYFILE"))

    ssl = case ssl_config do
            [] -> system_env("KASTLEX_KAFKA_USE_SSL", false, &s2b/1)
            kw -> kw
          end
    producer_config =
    []
      |> maybe_put(:required_acks, system_env("KASTLEX_PRODUCER_REQUIRED_ACKS", nil, &s2i/1))
      |> maybe_put(:ack_timeout, system_env("KASTLEX_PRODUCER_ACK_TIMEOUT", nil, &s2i/1))
      |> maybe_put(:partition_buffer_limit, system_env("KASTLEX_PRODUCER_PARTITION_BUFFER_LIMIT", nil, &s2i/1))
      |> maybe_put(:partition_onwire_limit, system_env("KASTLEX_PRODUCER_PARTITION_ONWIRE_LIMIT", nil, &s2i/1))
      |> maybe_put(:max_batch_size, system_env("KASTLEX_PRODUCER_MAX_BATCH_SIZE", nil, &s2i/1))
      |> maybe_put(:max_retries, system_env("KASTLEX_PRODUCER_MAX_RETRIES", nil, &s2i/1))
      |> maybe_put(:retry_backoff_ms, system_env("KASTLEX_PRODUCER_RETRY_BACKOFF_MS", nil, &s2i/1))
      |> maybe_put(:max_linger_ms, system_env("KASTLEX_PRODUCER_MAX_LINGER_MS", nil, &s2i/1))
      |> maybe_put(:max_linger_count, system_env("KASTLEX_PRODUCER_MAX_LINGER_COUNT", nil, &s2i/1))

    log.("brod producer config: #{inspect producer_config}")
    log.("brod ssl config: #{inspect ssl}")

    sasl = get_brod_sasl_config(system_env("KASTLEX_KAFKA_SASL_FILE"))
    case sasl do
      {mechanism, username, _password} ->
        log.("sasl #{mechanism} username: #{username}")
      _ ->
        log.("not using sasl auth")
    end

    [allow_topic_auto_creation: false,
     auto_start_producers: true,
     default_producer_config: producer_config,
     ssl: ssl,
     sasl: sasl]
  end

  # Example file format (yaml):
  # username: "foo",
  # password: "bar"
  # mechanism: "plain" # or scram_sha_256 or scram_sha_512
  defp get_brod_sasl_config(nil), do: :undefined
  defp get_brod_sasl_config(file) do
    try do
      {:ok, config} = YamlElixir.read_from_file(file)
      {sasl_mechanism(config), config["username"], config["password"]}
    rescue
      e in RuntimeError ->
        Logger.error("Error loading sasl config file: " <> e.message)
        :undefined
    catch
      _code, value ->
        Logger.error("Error loading sasl config file: #{inspect value}")
        :undefined
    end
  end

  defp sasl_mechanism(config) do
    case config["mechanism"] do
      nil             -> :plain
      "plain"         -> :plain
      "scram_sha_256" -> :scram_sha_256
      "scram_sha_512" -> :scram_sha_512
    end
  end

  defp maybe_put(kw, _key, nil),  do: kw
  defp maybe_put(kw, key, value), do: Keyword.put(kw, key, value)

  defp system_env(variable) do
    system_env(variable, nil)
  end

  defp system_env(variable, default) do
    system_env(variable, default, fn(x) -> x end)
  end

  defp system_env(variable, default, convert) do
    case System.get_env(variable) do
      nil -> default
      ""  -> default
      value -> convert.(value)
    end
  end

  defp maybe_log_parameter(_desc, nil), do: :ok
  defp maybe_log_parameter(desc, variable) do
    Logger.info "#{desc}: #{variable}"
  end

  defp parse_endpoints(endpoints) do
    endpoints
      |> String.split(",")
      |> Enum.map(&String.split(&1, ":"))
      |> Enum.map(fn([host, port]) -> {:erlang.binary_to_list(host),
                                       :erlang.binary_to_integer(port)} end)
  end

  defp s2b("true"), do: true
  defp s2b("1"),    do: true
  defp s2b(_),      do: false

  defp s2i(s), do: :erlang.binary_to_integer(s)

end
